cpp/src/neighbors/scann/detail/scann_build.cuh

```diff
@@ -0,0 +1,375 @@
+ */
+
+#pragma once
+
+#include "../../../cluster/kmeans_balanced.cuh"
+#include "../../detail/ann_utils.cuh"
+#include <cuvs/cluster/kmeans.hpp>
+#include <cuvs/neighbors/scann.hpp>
+
+#include <cub/cub.cuh>
+#include <cuda_bf16.h>
+#include <nvtx3/nvtx3.hpp>
+#include <raft/cluster/kmeans.cuh>
+#include <raft/cluster/kmeans_types.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/operators.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/linalg/add.cuh>
+#include <raft/linalg/coalesced_reduction.cuh>
+#include <raft/linalg/dot.cuh>
+#include <raft/linalg/gemm.hpp>
+#include <raft/linalg/gemv.cuh>
+#include <raft/linalg/linalg_types.hpp>
+#include <raft/linalg/map.cuh>
+#include <raft/linalg/matrix_vector.cuh>
+#include <raft/linalg/matrix_vector_op.cuh>
+#include <raft/linalg/multiply.cuh>
+#include <raft/linalg/normalize.cuh>
+#include <raft/linalg/power.cuh>
+#include <raft/linalg/reduce.cuh>
+#include <raft/linalg/transpose.cuh>
+#include <raft/matrix/argmin.cuh>
+#include <raft/matrix/copy.cuh>
+#include <raft/matrix/diagonal.cuh>
+#include <raft/matrix/init.cuh>
+#include <raft/matrix/slice.cuh>
+#include <raft/random/make_blobs.cuh>
+#include <raft/random/rng.cuh>
+#include <raft/sparse/neighbors/cross_component_nn.cuh>
+#include <raft/util/memory_type_dispatcher.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/sequence.h>
+#include <thrust/unique.h>
+
+#include <cuvs/distance/distance.hpp>
+
+#include "scann_avq.cuh"
+#include "scann_common.cuh"
+#include "scann_quantize.cuh"
+#include "scann_soar.cuh"
+
+#include <chrono>
+#include <cstdio>
+#include <vector>
+
+namespace cuvs::neighbors::experimental::scann::detail {
+using namespace cuvs::spatial::knn::detail;  // NOLINT
+
+/* @defgroup scann_build_detail scann build
+ * @{
+ */
+
+static const std::string RAFT_NAME = "raft";
+
+template <typename T,
+          typename IdxT     = int64_t,
+          typename Accessor = raft::host_device_accessor<std::experimental::default_accessor<T>,
+                                                         raft::memory_type::host>>
+index<T, IdxT> build(
+  raft::resources const& res,
+  const index_params& params,
+  raft::mdspan<const T, raft::matrix_extent<IdxT>, raft::row_major, Accessor> dataset)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(res);
+  IdxT dim            = dataset.extent(1);
+
+  RAFT_LOG_DEBUG("Creating empty index");
+
+  index<T, IdxT> idx(res, params, dataset.extent(0), dataset.extent(1));
+  raft::device_matrix_view<T, int64_t> centroids_view = idx.centers();
+  raft::random::RngState random_state{137};
+  // sample/project
+
+  cuvs::cluster::kmeans::balanced_params kmeans_params;
+  kmeans_params.n_iters = params.kmeans_n_iters;
+
+  {
+    auto trainset =
```

Comment:
How large this matrix typically is? If there is a chance to go out of memory, then it would be great throw a descriptive error message, e.g.

https://github.com/rapidsai/cuvs/blob/ed32f0fb6689de6a346f7d551662d58e2965588e/cpp/src/neighbors/ivf_pq/ivf_pq_build.cuh#L1738-L1741